### PR TITLE
fix: no gaps in root btc source

### PR DESCRIPTION
**Pull request by kylezs**
# Pull Request

## Checklist

Please conduct a thorough self-review before opening the PR.

- [x] I am confident that the code works.
- [x] I have written sufficient tests.
- [x] I have written and tested required migrations.
- [x] I have updated documentation where appropriate.

## Summary

Fixes an extreme BTC witnessing edge case that's only currently possible when a sub 5 second bitcoin block occurs. In most cases this is handled. But can occur under very specific rotation conditions. This will be more simply resolved with the upcoming larger witnessing refactor.

### CODE REVIEW

File: engine/src/witness/common/chain_source/strictly_monotonic.rs [hunk 1]
```diff
@@ -83,14 +83,15 @@ mod test {
 					Header { index: 4u32, hash: (), parent_hash: Some(()), data: () },
 					Header { index: 3u32, hash: (), parent_hash: Some(()), data: () },
 					Header { index: 2u32, hash: (), parent_hash: Some(()), data: () },
+					Header { index: 10u32, hash: (), parent_hash: Some(()), data: () },
```
Comment (dandanlen) [inline engine/src/witness/common/chain_source/strictly_monotonic.rs:86]: Is the addition of this to test case meaningful? How? (Would it have previously not passed?)

File: engine/src/witness/btc/source.rs [hunk 2]
```diff
@@ -40,22 +47,88 @@ where
...
+						}
+
 						tick.tick().await;
 
 						let best_block_header = client.best_block_header().await;
-						if last_block_hash_yielded != Some(best_block_header.hash) {
+
+						let yield_new_best_header: bool = match &mut stream_state {
+							Some(state)
+								if state.last_block_yielded_index.saturating_add(1) ==
```
Comment (MxmUrw) [inline engine/src/witness/btc/source.rs:92]: What purpose does the `state.last_block_yielded_index.saturating_add(1) == best_block_header.height` serve? It seems that the hash check should be enough, no?

File: engine/src/witness/common/chain_source/strictly_monotonic.rs (hunk 1 above)
Comment (kylezs) [reply engine/src/witness/common/chain_source/strictly_monotonic.rs:86]: It's not meaningful, it's just not obvious that it allows skipping blocks without that. Which informed me a little on how the rest of this would/could work

File: engine/src/witness/btc/source.rs [hunk 3]
```diff
@@ -40,22 +47,90 @@ where
...
+							// yield the new best header immediately
+							None => true,
+							// If we've progressed by more than one block, then we need to backfill
+							Some(state)
+								if state.last_block_yielded_index < best_block_header.height =>
+							{
+								// Update the state for the next iteration to backfill
+								if let Some(state) = stream_state.as_mut() {
+									state.best_known_block_index = best_block_header.height;
+								}
```
Comment (MxmUrw) [inline engine/src/witness/btc/source.rs:108]: We already have a mutable reference to `state`, why are we creating a new one? I'm actually surprised that this works, because it looks like we have two mut references to the same thing?

**Review by MxmUrw — APPROVED** [review]
LGTM!
